File: design-pattern-jvm/app/src/main/java/design/pattern/java/patterns/behavioral/state/TrafficLight.java
package design.pattern.java.patterns.behavioral.state;

public class TrafficLight {
  private TrafficLightState state;
  public TrafficLight(TrafficLightState initialState) {
    this.state = initialState == null ? new RedLight() : initialState;
  }
  public TrafficLight changeLight() {
    this.state = this.state.changeLight();
    return this;
  }
  public String trafficAction() {
    return this.state.trafficAction();
  }
  public TrafficLightState getCurrentState() {
    return this.state;
  }
}

File: design-pattern-jvm/app/src/main/kotlin/design/pattern/kotlin/patterns/behavioral/state/TrafficLight.kt
package design.pattern.kotlin.patterns.behavioral.state

class TrafficLight(private var state: TrafficLightState = RedLight()) {
  fun changeLight(): TrafficLight {
    this.state = this.state.changeLight()
    return this
  }
  fun trafficAction(): String {
    return this.state.trafficAction()
  }
  fun currentState(): TrafficLightState {
    return this.state
  }
}
